### PR TITLE
L016: New option "ignore_comment_clause" to ignore column COMMENTs, etc.

### LINKED ISSUE #2554: Update L016 to optionally ignore long COMMENT lines as well
For example:
```
CREATE OR REPLACE TABLE mytable (
    col1 NUMBER COMMENT 'col1 comment',
    col2 BOOLEAN COMMENT 'very long comment line, exceeding max_line_length...',
    col3 NUMBER COMMENT 'col3 comment'
)

```
sqlfluff-fix throws `L016 | Line is too long` error assuming the col2 comment exceeds max_line_length setting.

Setting `ignore_comment_lines` to True also doesn't help.
Updating each of the offending lines with --noqa directive might not be scalable where there are 100s of files involved.

### PR DESCRIPTION
**Pull request by barrywhart**
### Brief summary of the change made
Fixes #2554

### Pull Request checklist
- [x] Please confirm you have completed any of the necessary steps below.

- Included test cases to demonstrate any code changes, which may be one or more of the following:
  - `.yml` rule test cases in `test/fixtures/rules/std_rule_cases`.
  - `.sql`/`.yml` parser test cases in `test/fixtures/dialects` (note YML files can be auto generated with `tox -e generate-fixture-yml`).
  - Full autofix test cases in `test/fixtures/linter/autofix`.
  - Other.
- Added appropriate documentation for the change.
- Created GitHub issues for any relevant followup/future enhancements if appropriate.

### CODE REVIEW

File: src/sqlfluff/core/default_config.cfg [hunk 1]
```diff
@@ -92,6 +92,7 @@ ignore_words = None
 [sqlfluff:rules:L016]
 # Line length
 ignore_comment_lines = False
+ignore_comment_clause = False
```
Comment (tunetheweb) [inline src/sqlfluff/core/default_config.cfg:95]: Make this consistent with config above

```suggestion
ignore_comment_clauses = False
```

File: src/sqlfluff/core/rules/config_info.py [hunk 2]
```diff
@@ -81,6 +81,13 @@
             " be ignored when linting line lengths?"
         ),
     },
+    "ignore_comment_clause": {
```
Comment (tunetheweb) [inline src/sqlfluff/core/rules/config_info.py:84]:
```suggestion
    "ignore_comment_clauses": {
```

File: src/sqlfluff/rules/L016.py [hunk 3]
```diff
@@ -29,6 +30,7 @@ class Rule_L016(Rule_L003):
         "tab_space_size",
         "indent_unit",
         "ignore_comment_lines",
+        "ignore_comment_clause",
```
Comment (tunetheweb) [inline src/sqlfluff/rules/L016.py:33]:
```suggestion
        "ignore_comment_clauses",
```

File: src/sqlfluff/rules/L016.py [hunk 4]
```diff
@@ -418,11 +420,15 @@ def _compute_segment_length(cls, segment: BaseSegment) -> int:
...
 
-    @classmethod
-    def _compute_source_length(cls, segments: Sequence[BaseSegment]) -> int:
+    def _compute_source_length(
+        self, segments: Sequence[BaseSegment], memory: dict
+    ) -> int:
         line_len = 0
         seen_slices = set()
         for segment in segments:
+            if self.ignore_comment_clause and segment in memory["comment_clauses"]:
```
Comment (tunetheweb) [inline src/sqlfluff/rules/L016.py:429]:
```suggestion
            if self.ignore_comment_clauses and segment in memory["comment_clauses"]:
```

File: src/sqlfluff/rules/L016.py [hunk 5]
```diff
@@ -458,16 +464,30 @@ def _eval(self, context: RuleContext) -> Optional[LintResult]:
         # Config type hints
         self.max_line_length: int
         self.ignore_comment_lines: bool
+        self.ignore_comment_clause: bool
```
Comment (tunetheweb) [inline src/sqlfluff/rules/L016.py:467]:
```suggestion
        self.ignore_comment_clauses: bool
```

File: src/sqlfluff/rules/L016.py [hunk 6]
```diff
@@ -458,16 +464,30 @@ def _eval(self, context: RuleContext) -> Optional[LintResult]:
...
 
+        if not context.memory:
+            memory: dict = {"comment_clauses": set()}
+        else:
+            memory = context.memory
         if context.segment.name == "newline":
             # iterate to buffer the whole line up to this point
             this_line = self._gen_line_so_far(context.raw_stack)
         else:
+            if self.ignore_comment_clause and context.segment.is_type(
```
Comment (tunetheweb) [inline src/sqlfluff/rules/L016.py:477]:
```suggestion
            if self.ignore_comment_clauses and context.segment.is_type(
```

File: test/fixtures/rules/std_rule_cases/L016.yml [hunk 7]
```diff
@@ -238,3 +238,55 @@ test_issue_1666_line_too_long_unfixable_jinja:
...
+
+test_pass_ignore_comment_clause_snowflake:
+  pass_str: |
+    CREATE TABLE foo_table (bar INTEGER) COMMENT = 'very long comment line, exceeding max_line_lengthvery long comment line, exceeding max_line_length very long comment line, exceeding max_line_length very long comment line, exceeding max_line_length'
+  configs:
+    core:
+      dialect: snowflake
+    rules:
+      L016:
+        ignore_comment_clause: True
```
Comment (tunetheweb) [inline test/fixtures/rules/std_rule_cases/L016.yml:292]: Should we add a test case for when the line is very long - even when ignoring the comment clause? It should still fail.

**Review by tunetheweb — CHANGES_REQUESTED** [review]
Pro github tip: to accept lots of changes (like I've suggested here) use the "files" tab and add all the suggestions to a batch and then commit that batch in one commit.

File: test/fixtures/rules/std_rule_cases/L016.yml (hunk 7 above)
Comment (barrywhart) [reply test/fixtures/rules/std_rule_cases/L016.yml:292]: Will do

Comment (barrywhart) [reply test/fixtures/rules/std_rule_cases/L016.yml:306]: Done

**Comment by barrywhart** [comment]
@tunetheweb: I added the test case you suggested. Ready for another review.

**Comment by tunetheweb** [comment]
The tests were failing so was waiting for them to pass before approving.

**Comment by barrywhart** [comment]
Hmm, maybe I broke something. Will look again in a bit.

**Comment by barrywhart** [comment]
@tunetheweb: I think it'll work now. We just needed to apply the same config setting rename to the test case `.yml` file as the code itself.

**Review by tunetheweb — APPROVED** [review]

**Pull request merged**